src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -81,37 +88,108 @@ private String decodeAlarm(String event) {
                 return Position.ALARM_GEOFENCE_EXIT;
             case "6":
                 return Position.ALARM_OVERSPEED;
+            case "5":
+                return Position.ALARM_POWER_CUT;
             case "3":
                 return Position.ALARM_SOS;
             default:
                 return null;
         }
     }
 
-    @Override
-    protected Object decode(
-            Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+    private String decodeEvent(String event, Position position) {
 
-        String sentence = (String) msg;
+        position.set(Position.KEY_IGNITION, false);
 
-        if (sentence.startsWith("$ECHK") && channel != null) {
-            channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress)); // heartbeat
-            return null;
+        if (event.length() == 1) {
+            char inputStatus = event.charAt(0);
+            if (inputStatus >= 'A' && inputStatus <= 'D') {
+                int inputStatusInt = inputStatus - 'A';
+                position.set(Position.PREFIX_IN + 1, inputStatusInt & 1);
+                position.set(Position.PREFIX_IN + 2, inputStatusInt & 2);
+                if ((inputStatusInt & 1) != 0) {
+                    position.set(Position.KEY_IGNITION, true);
+                }
+                return null;
+            }
+        }
+
+        return event;
+    }
+
+    private void sendEventResponse(
+            String event, String devicePassword, Channel channel, SocketAddress remoteAddress) {
+
+        String responseCode = null;
+
+        switch (event) {
+            case "3":
+                responseCode = "d";
+                break;
+            case "S":
+            case "T":
+                responseCode = "t";
+                break;
+            case "X":
+            case "4":
+                responseCode = "x";
+                break;
+            case "Y":
+                responseCode = "y";
+                break;
+            case "Z":
+                responseCode = "z";
+                break;
+            default:
+                break;
+        }
+
+        if (responseCode != null) {
+            String response = "$AVCFG," + devicePassword + "," + responseCode;
+            response += Checksum.nmea(response) + "\r\n";
+            channel.writeAndFlush(new NetworkMessage(response, remoteAddress));
+        }
+
+    }
+
+    private void sendAcknowledge(
+            String status, String event, String checksum, Channel channel, SocketAddress remoteAddress) {
+
+        if (Character.isLowerCase(status.charAt(0))) {
+            String response = "$EAVACK," + event + "," + checksum;
+            response += Checksum.nmea(response) + "\r\n";
+            channel.writeAndFlush(new NetworkMessage(response, remoteAddress));
+        }
+
+    }
+
+    private Object decodeEchk(
```

Comment:
OK, I think now it doesn't make sense to have this method. All it does is this:
```java
if (channel != null) {
    channel.writeAndFlush(new NetworkMessage(sentence + "\r\n", remoteAddress));
}
```

Reply:
Again, just trying to keep the code uniform. No problem, removing it. Thanks...